File: frontend/src/api/canvas.ts
import { apiClient } from './client'

export type CanvasNodeBase = {
  id: string
  x: number
  y: number
  w: number
  h: number
}

export type CanvasQuestionNode = CanvasNodeBase & {
  kind: 'question'
  questionId: string
  title: string
  stemHtml?: string
  meta?: Record<string, unknown>
  type?: string
  difficulty?: string
  knowledgePoints?: string[]
  source?: string
  url?: string
  selectReason?: string
}

export type CanvasNoteNode = CanvasNodeBase & {
  kind: 'note'
  text: string
}

export type CanvasNode = CanvasQuestionNode | CanvasNoteNode

export type CanvasSnapshot = {
  version: 1
  nodes: CanvasNode[]
}

export interface CanvasBoardSummary {
  id: number
  title: string
  subject: string
  revision: number
  createdAt: string
  updatedAt: string
}

export interface CanvasBoard extends CanvasBoardSummary {
  snapshot: CanvasSnapshot
}

export interface CanvasBoardVersion {
  id: number
  boardId: number
  revision: number
  createdAt: string
  snapshot?: CanvasSnapshot
}

export interface PickedCanvasQuestion {
  success: boolean
  questionId: string
  title: string
  meta?: Record<string, unknown>
  stemHtml: string
  type?: string
  difficulty?: string
  knowledgePoints?: string[]
  source?: string
  url?: string
  selectReason?: string
}

export type UpdateCanvasBoardResult =
  | { success: true; board: CanvasBoard; conflict?: false }
  | { success: false; conflict: true; serverBoard: CanvasBoard; error?: string }

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object')
}

function defaultSnapshot(): CanvasSnapshot {
  return { version: 1, nodes: [] }
}

function toStringValue(value: unknown): string {
  return typeof value === 'string' ? value : ''
}

function toNumberValue(value: unknown, fallback = 0): number {
  const num = Number(value)
  return Number.isFinite(num) ? num : fallback
}

function normalizeSnapshot(value: unknown): CanvasSnapshot {
  if (!isRecord(value)) return defaultSnapshot()
  const nodes = Array.isArray(value.nodes) ? value.nodes : []
  return {
    version: 1,
    nodes: nodes.filter(isRecord).map((node) => normalizeNode(node)).filter(Boolean) as CanvasNode[],
  }
}

function normalizeNode(node: Record<string, unknown>): CanvasNode | null {
  const kind = node.kind === 'question' ? 'question' : node.kind === 'note' ? 'note' : ''
  if (!kind) return null
  const base = {
    id: toStringValue(node.id) || `${kind}-${Math.random().toString(36).slice(2, 10)}`,
    x: toNumberValue(node.x),
    y: toNumberValue(node.y),
    w: Math.max(160, toNumberValue(node.w, kind === 'question' ? 340 : 260)),
    h: Math.max(120, toNumberValue(node.h, kind === 'question' ? 240 : 160)),
  }
  if (kind === 'note') {
    return { ...base, kind, text: toStringValue(node.text) }
  }
  return {
    ...base,
    kind,
    questionId: toStringValue(node.questionId || node.question_id),
    title: toStringValue(node.title) || '题目',
    stemHtml: toStringValue(node.stemHtml || node.stem_html),
    meta: isRecord(node.meta) ? node.meta : undefined,
    type: toStringValue(node.type),
    difficulty: toStringValue(node.difficulty),
    knowledgePoints: Array.isArray(node.knowledgePoints)
      ? node.knowledgePoints.map(String)
      : Array.isArray(node.knowledge_points)
        ? node.knowledge_points.map(String)
        : undefined,
    source: toStringValue(node.source),
    url: toStringValue(node.url),
    selectReason: toStringValue(node.selectReason || node.select_reason),
  }
}

function normalizeSummary(raw: unknown): CanvasBoardSummary {
  const item = isRecord(raw) ? raw : {}
  return {
    id: toNumberValue(item.id),
    title: toStringValue(item.title) || '未命名画布',
    subject: toStringValue(item.subject),
    revision: toNumberValue(item.revision, 1),
    createdAt: toStringValue(item.createdAt || item.created_at),
    updatedAt: toStringValue(item.updatedAt || item.updated_at),
  }
}

function normalizeBoard(raw: unknown): CanvasBoard {
  const item = isRecord(raw) ? raw : {}
  return {
    ...normalizeSummary(item),
    snapshot: normalizeSnapshot(item.snapshot),
  }
}

function normalizeVersion(raw: unknown): CanvasBoardVersion {
  const item = isRecord(raw) ? raw : {}
  const version: CanvasBoardVersion = {
    id: toNumberValue(item.id),
    boardId: toNumberValue(item.boardId || item.board_id),
    revision: toNumberValue(item.revision),
    createdAt: toStringValue(item.createdAt || item.created_at),
  }
  if ('snapshot' in item) version.snapshot = normalizeSnapshot(item.snapshot)
  return version
}

function ensureSuccessEnvelope(data: unknown, key: string): Record<string, unknown> {
  if (!isRecord(data) || data.success === false) {
    const message = isRecord(data) ? toStringValue(data.error) || 'canvas_request_failed' : 'canvas_request_failed'
    throw new Error(message)
  }
  if (!(key in data)) throw new Error('canvas_response_missing_payload')
  return data
}

export async function getCanvasBoards(params?: { q?: string; limit?: number }): Promise<CanvasBoardSummary[]> {
  const response = await apiClient.get<unknown>('/canvas/boards', {
    params: {
      q: params?.q,
      limit: params?.limit,
    },
  })
  const data = ensureSuccessEnvelope(response.data, 'boards')
  return Array.isArray(data.boards) ? data.boards.map(normalizeSummary) : []
}

export async function getCanvasBoard(id: number | string): Promise<CanvasBoard> {
  const response = await apiClient.get<unknown>(`/canvas/boards/${id}`)
  const data = ensureSuccessEnvelope(response.data, 'board')
  return normalizeBoard(data.board)
}

export async function createCanvasBoard(data: {
  title?: string
  subject?: string
  snapshot?: CanvasSnapshot
}): Promise<CanvasBoard> {
  const response = await apiClient.post<unknown>('/canvas/boards', data)
  const payload = ensureSuccessEnvelope(response.data, 'board')
  return normalizeBoard(payload.board)
}

export async function updateCanvasBoard(
  id: number | string,
  data: {
    title?: string
    subject?: string
    snapshot?: CanvasSnapshot
    expectedRevision?: number
  }
): Promise<UpdateCanvasBoardResult> {
  const response = await apiClient.put<unknown>(`/canvas/boards/${id}`, {
    title: data.title,
    subject: data.subject,
    snapshot: data.snapshot,
    expected_revision: data.expectedRevision,
  })
  const payload = response.data
  if (isRecord(payload) && payload.success === false && payload.conflict === true) {
    return {
      success: false,
      conflict: true,
      serverBoard: normalizeBoard(payload.server_board),
      error: toStringValue(payload.error),
    }
  }
  const ok = ensureSuccessEnvelope(payload, 'board')
  return { success: true, board: normalizeBoard(ok.board) }
}

export async function getCanvasBoardVersions(boardId: number | string, limit = 30): Promise<CanvasBoardVersion[]> {
  const response = await apiClient.get<unknown>(`/canvas/boards/${boardId}/versions`, { params: { limit } })
  const payload = ensureSuccessEnvelope(response.data, 'versions')
  return Array.isArray(payload.versions) ? payload.versions.map(normalizeVersion) : []
}

export async function createCanvasBoardVersion(boardId: number | string): Promise<{
  version: CanvasBoardVersion
  versions: CanvasBoardVersion[]
}> {
  const response = await apiClient.post<unknown>(`/canvas/boards/${boardId}/versions`)
  const payload = ensureSuccessEnvelope(response.data, 'version')
  return {
    version: normalizeVersion(payload.version),
    versions: Array.isArray(payload.versions) ? payload.versions.map(normalizeVersion) : [],
  }
}

export async function getCanvasBoardVersion(boardId: number | string, versionId: number | string): Promise<CanvasBoardVersion> {
  const response = await apiClient.get<unknown>(`/canvas/boards/${boardId}/versions/${versionId}`)
  const payload = ensureSuccessEnvelope(response.data, 'version')
  return normalizeVersion(payload.version)
}

export async function pickCanvasQuestions(
  boardId: number | string,
  payload: {
    requirement: string
    subject?: string
    eduLevel?: string
    count?: number
    limit?: number
    maxPages?: number
  }
): Promise<{ selectedIds: string[]; questions: PickedCanvasQuestion[] }> {
  const response = await apiClient.post<unknown>(`/canvas/boards/${boardId}/pick-questions`, {
    requirement: payload.requirement,
    subject: payload.subject,
    edu_level: payload.eduLevel,
    count: payload.count,
    limit: payload.limit,
    max_pages: payload.maxPages,
  })
  const data = ensureSuccessEnvelope(response.data, 'questions')
  return {
    selectedIds: Array.isArray(data.selected_ids) ? data.selected_ids.map(String) : [],
    questions: Array.isArray(data.questions)
      ? data.questions.filter(isRecord).map((question) => ({
          success: question.success !== false,
          questionId: toStringValue(question.question_id || question.questionId),
          title: toStringValue(question.title) || '题目',
          meta: isRecord(question.meta) ? question.meta : undefined,
          stemHtml: toStringValue(question.stem_html || question.stemHtml),
          type: toStringValue(question.type),
          difficulty: toStringValue(question.difficulty),
          knowledgePoints: Array.isArray(question.knowledge_points)
            ? question.knowledge_points.map(String)
            : Array.isArray(question.knowledgePoints)
              ? question.knowledgePoints.map(String)
              : undefined,
          source: toStringValue(question.source),
          url: toStringValue(question.url),
          selectReason: toStringValue(question.select_reason || question.selectReason),
        }))
      : [],
  }
}

export async function renderCanvasQuestion(
  questionId: string,
  params?: { subject?: string; eduLevel?: string }
): Promise<PickedCanvasQuestion> {
  const response = await apiClient.get<unknown>(`/canvas/questions/${encodeURIComponent(questionId)}/render`, {
    params: {
      subject: params?.subject,
      edu_level: params?.eduLevel,
    },
  })
  const data = ensureSuccessEnvelope(response.data, 'question')
  const question = isRecord(data.question) ? data.question : {}
  return {
    success: true,
    questionId: toStringValue(question.question_id || question.questionId || questionId),
    title: toStringValue(question.title) || '题目',
    meta: isRecord(question.meta) ? question.meta : undefined,
    stemHtml: toStringValue(question.stem_html || question.stemHtml),
    type: toStringValue(question.type),
    difficulty: toStringValue(question.difficulty),
    knowledgePoints: Array.isArray(question.knowledge_points)
      ? question.knowledge_points.map(String)
      : Array.isArray(question.knowledgePoints)
        ? question.knowledgePoints.map(String)
        : undefined,
    source: toStringValue(question.source),
    url: toStringValue(question.url),
    selectReason: toStringValue(question.select_reason || question.selectReason),
  }
}

File: backend/api/canvas.py
from __future__ import annotations

import asyncio
import json
import os
from typing import Any, Dict, List, Optional
from urllib.parse import quote, urlparse

from bs4 import BeautifulSoup
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from backend.api.auth import require_auth
from backend.api.canvas_schemas import CanvasBoardCreate, CanvasBoardUpdate
from backend.config import DEFAULT_SUBJECT
from backend.crawler_manager import get_crawler
from backend.database.models import (
    create_canvas_board,
    create_canvas_board_version,
    get_canvas_board,
    get_canvas_board_version,
    list_canvas_board_versions,
    list_canvas_boards,
    update_canvas_board,
)
from backend.subjects import resolve_subject

router = APIRouter(prefix="/canvas", dependencies=[Depends(require_auth)])


def _parse_snapshot(raw: str) -> Dict[str, Any]:
    raw = (raw or "").strip()
    if not raw:
        return {}
    try:
        parsed = json.loads(raw)
    except Exception:
        return {}
    return parsed if isinstance(parsed, dict) else {}


def _snapshot_max_bytes() -> int:
    raw = str(os.getenv("CANVAS_SNAPSHOT_MAX_BYTES") or "").strip()
    try:
        value = int(raw) if raw else 2 * 1024 * 1024
    except Exception:
        value = 2 * 1024 * 1024
    return max(1024, min(value, 20 * 1024 * 1024))


def _serialize_snapshot(snapshot: Dict[str, Any]) -> str:
    try:
        raw = json.dumps(snapshot, ensure_ascii=False)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=f"invalid_snapshot: {str(exc)}")

    if len(raw.encode("utf-8")) > _snapshot_max_bytes():
        raise HTTPException(status_code=413, detail="snapshot_too_large")
    return raw


def _normalize_asset_url(raw: str, base_url: str) -> str:
    raw = (raw or "").strip()
    if not raw:
        return ""
    if raw.startswith("//"):
        return f"https:{raw}"
    if raw.startswith("/"):
        return f"{base_url.rstrip('/')}{raw}"
    return raw


def _is_safe_http_url(url: str) -> bool:
    url = (url or "").strip()
    if not url:
        return False
    parsed = urlparse(url)
    if parsed.scheme not in {"http", "https"}:
        return False
    return bool(parsed.netloc)


def _sanitize_question_html(html: str, *, base_url: str) -> str:
    """
    Sanitize and rewrite question stem HTML for safe rendering in the frontend.

    - Removes potentially dangerous tags.
    - Strips inline event handlers (onClick, etc.).
    - Rewrites <img> sources to `/api/media/proxy?url=...` for on-demand download + cache.
    """
    html = html or ""
    soup = BeautifulSoup(html, "lxml")

    for tag in soup(["script", "style", "iframe", "object", "embed", "link", "meta"]):
        tag.decompose()

    for tag in soup.find_all(True):
        for attr in list(tag.attrs):
            if attr.lower().startswith("on"):
                del tag.attrs[attr]

    for img in soup.find_all("img"):
        src = (img.get("src") or img.get("data-src") or "").strip()
        if not src:
            img.decompose()
            continue
        full = _normalize_asset_url(src, base_url)
        if not full or not _is_safe_http_url(full):
            img.decompose()
            continue
        img.attrs = {"src": f"/api/media/proxy?url={quote(full, safe='')}", "loading": "lazy"}

    for a in soup.find_all("a"):
        href = (a.get("href") or "").strip()
        if not href:
            continue
        full = _normalize_asset_url(href, base_url)
        if not full or not _is_safe_http_url(full):
            # Keep the text, but prevent scriptable/unsafe links.
            if "href" in a.attrs:
                del a.attrs["href"]
            if "target" in a.attrs:
                del a.attrs["target"]
            if "rel" in a.attrs:
                del a.attrs["rel"]
            continue
        a.attrs = {
            **{k: v for k, v in a.attrs.items() if k not in {"href", "target", "rel"}},
            "href": full,
            "target": "_blank",
            "rel": "noreferrer noopener",
        }

    container = soup.body if soup.body else soup
    return container.decode_contents()


class CanvasPickQuestionsRequest(BaseModel):
    requirement: str = Field(min_length=1, max_length=800)
    subject: Optional[str] = Field(default=None, max_length=100)
    edu_level: Optional[str] = Field(default=None, max_length=50)
    count: int = Field(default=3, ge=1, le=10)
    model: Optional[str] = Field(default=None, max_length=200)
    limit: int = Field(default=25, ge=5, le=50)
    max_pages: int = Field(default=2, ge=1, le=8)


@router.get("/boards")
async def list_boards(limit: int = Query(50, ge=1, le=200), q: str = Query(""), user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    boards = await list_canvas_boards(user_id=user_id, limit=limit, query=q)
    return {"success": True, "boards": boards}


@router.post("/boards")
async def create_board(payload: CanvasBoardCreate, user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    snapshot_raw = ""
    if payload.snapshot is not None:
        snapshot_raw = _serialize_snapshot(payload.snapshot)

    board = await create_canvas_board(
        user_id=user_id,
        title=(payload.title or "").strip(),
        subject=(payload.subject or "").strip(),
        snapshot=snapshot_raw,
    )
    return {"success": True, "board": board}


@router.get("/boards/{board_id}")
async def get_board(board_id: int, user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    board = await get_canvas_board(user_id=user_id, board_id=board_id)
    if not board:
        raise HTTPException(status_code=404, detail="board_not_found")

    snapshot = _parse_snapshot(board.get("snapshot", ""))
    out = dict(board)
    out["snapshot"] = snapshot
    return {"success": True, "board": out}


@router.put("/boards/{board_id}")
async def put_board(board_id: int, payload: CanvasBoardUpdate, user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    snapshot_raw = None
    if payload.snapshot is not None:
        snapshot_raw = _serialize_snapshot(payload.snapshot)

    result = await update_canvas_board(
        user_id,
        board_id,
        title=payload.title,
        subject=payload.subject,
        snapshot=snapshot_raw,
        expected_revision=payload.expected_revision,
    )

    if not result.get("success") and result.get("conflict"):
        server_board = result.get("server_board") or {}
        server_board["snapshot"] = _parse_snapshot(server_board.get("snapshot", ""))
        return {"success": False, "conflict": True, "server_board": server_board}

    if not result.get("success"):
        raise HTTPException(status_code=400, detail=result.get("error") or "update_failed")

    board = result.get("board") or {}
    board["snapshot"] = _parse_snapshot(board.get("snapshot", ""))
    return {"success": True, "board": board}


@router.get("/boards/{board_id}/versions")
async def get_versions(board_id: int, limit: int = Query(30, ge=1, le=200), user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    versions = await list_canvas_board_versions(user_id=user_id, board_id=board_id, limit=limit)
    return {"success": True, "versions": versions}


@router.post("/boards/{board_id}/versions")
async def create_version(board_id: int, user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    result = await create_canvas_board_version(user_id=user_id, board_id=board_id)
    if not result.get("success"):
        raise HTTPException(status_code=404, detail=result.get("error") or "board_not_found")
    return result


@router.get("/boards/{board_id}/versions/{version_id}")
async def get_version(board_id: int, version_id: int, user: dict = Depends(require_auth)) -> dict:
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    version = await get_canvas_board_version(user_id=user_id, board_id=board_id, version_id=version_id)
    if not version:
        raise HTTPException(status_code=404, detail="version_not_found")
    snapshot = _parse_snapshot(version.get("snapshot", ""))
    out = dict(version)
    out["snapshot"] = snapshot
    return {"success": True, "version": out}


@router.post("/boards/{board_id}/pick-questions")
async def pick_questions(board_id: int, payload: CanvasPickQuestionsRequest, user: dict = Depends(require_auth)) -> dict:
    """
    Use crawler + MCP sub-AI selector to pick questions, then return render-ready HTML.

    Notes:
    - Formula rendering: inline SVG (no svg2latex).
    - Images: rewritten to `/api/media/proxy` for on-demand download + cache.
    """
    user_id = str((user or {}).get("user_id") or "").strip() or "1"
    board = await get_canvas_board(user_id=user_id, board_id=board_id)
    if not board:
        raise HTTPException(status_code=404, detail="board_not_found")

    requirement = (payload.requirement or "").strip()
    if not requirement:
        raise HTTPException(status_code=400, detail="requirement_required")

    subject_input = (payload.subject or board.get("subject") or DEFAULT_SUBJECT).strip()
    edu_level = (payload.edu_level or "").strip()

    try:
        subject = resolve_subject(subject_input, edu_level=edu_level, strict=True)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    crawler = await get_crawler(subject=subject, edu_level=edu_level, strict=True)

    search_result = await crawler.search_by_keyword(
        keyword=requirement,
        edu_level=edu_level,
        limit=max(int(payload.limit), int(payload.count) * 6),
        max_pages=int(payload.max_pages),
        require_difficulty=True,
        strict_subject=True,
    )
    if not search_result.get("success"):
        return {"success": False, "error": "search_failed", "details": search_result}

    candidates = search_result.get("questions") or []
    # Stable de-dupe while preserving order
    seen_ids = set()
    candidate_questions: List[Dict[str, Any]] = []
    for q in candidates:
        if not isinstance(q, dict):
            continue
        qid = (q.get("question_id") or "").strip()
        if not qid or qid in seen_ids:
            continue
        seen_ids.add(qid)
        candidate_questions.append(q)

    if not candidate_questions:
        return {
            "success": False,
            "error": "no_candidates",
            "details": {"subject": subject, "edu_level": edu_level, "search_result": search_result},
        }

    # MCP sub-AI selector (picks one from 2-8 candidates each round)
    from backend.mcp.sub_ai_selector import select_best_question

    selected_ids: List[str] = []
    selection: List[Dict[str, Any]] = []
    remaining = list(candidate_questions)

    for _ in range(int(payload.count)):
        if not remaining:
            break

        subset = remaining[: min(len(remaining), 8)]
        if len(subset) == 1:
            chosen_id = str(subset[0].get("question_id") or "").strip()
            if chosen_id:
                selected_ids.append(chosen_id)
                selection.append(
                    {
                        "selected_question_id": chosen_id,
                        "reason": "候选不足，直接选取",
                        "analysis": "",
                    }
                )
                remaining = [q for q in remaining if (q.get("question_id") or "").strip() != chosen_id]
            continue

        req = requirement
        if selected_ids:
            req = f"{requirement}\n已选题目ID：{', '.join(selected_ids)}。请不要重复选择已选题。"

        sel = await select_best_question(questions=subset, requirement=req, model=payload.model)
        if not sel.get("success"):
            return {"success": False, "error": "select_failed", "details": sel}

        chosen_id = str(sel.get("selected_question_id") or "").strip()
        if not chosen_id:
            return {"success": False, "error": "select_failed_empty_id", "details": sel}

        selection.append(sel)
        selected_ids.append(chosen_id)
        remaining = [q for q in remaining if (q.get("question_id") or "").strip() != chosen_id]

    if not selected_ids:
        return {"success": False, "error": "select_failed_no_results", "details": selection}

    tasks = [
        crawler.get_question_detail(qid, formula_mode="svg", stem_mode="html") for qid in selected_ids
    ]
    details = await asyncio.gather(*tasks, return_exceptions=True)

    reason_by_id = {
        str(s.get("selected_question_id") or "").strip(): (s.get("reason") or "")
        for s in selection
        if isinstance(s, dict)
    }

    out_questions: List[Dict[str, Any]] = []
    for qid, detail in zip(selected_ids, details):
        if isinstance(detail, Exception):
            out_questions.append(
                {
                    "question_id": qid,
                    "success": False,
                    "error": str(detail),
                    "select_reason": reason_by_id.get(qid, ""),
                }
            )
            continue

        if not isinstance(detail, dict) or not detail.get("success"):
            out_questions.append(
                {
                    "question_id": qid,
                    "success": False,
                    "error": (detail or {}).get("error") if isinstance(detail, dict) else "detail_failed",
                    "detail": detail,
                    "select_reason": reason_by_id.get(qid, ""),
                }
            )
            continue

        stem_html = _sanitize_question_html(detail.get("stem_html", ""), base_url=getattr(crawler, "base_url", ""))
        meta_parts = []
        if (detail.get("type") or "").strip():
            meta_parts.append(str(detail.get("type")).strip())
        if (detail.get("difficulty") or "").strip():
            meta_parts.append(str(detail.get("difficulty")).strip())
        meta = " · ".join(meta_parts)

        out_questions.append(
            {
                "success": True,
                "question_id": detail.get("question_id") or qid,
                "title": f"题目 {detail.get('question_id') or qid}",
                "meta": meta,
                "stem_html": stem_html,
                "type": detail.get("type") or "",
                "difficulty": detail.get("difficulty") or "",
                "knowledge_points": detail.get("knowledge_points") or "",
                "source": detail.get("source") or "",
                "url": detail.get("url") or "",
                "select_reason": reason_by_id.get(qid, ""),
            }
        )

    return {
        "success": True,
        "subject": subject,
        "edu_level": edu_level,
        "requirement": requirement,
        "candidate_count": len(candidate_questions),
        "selected_ids": selected_ids,
        "questions": out_questions,
        "selection": selection,
    }


@router.get("/questions/{question_id}/render")
async def render_question(
    question_id: str,
    subject: str = Query("", max_length=100),
    edu_level: str = Query("", max_length=50),
) -> dict:
    """Render a single question as HTML with inline SVG formulas and cached images."""
    subject_input = (subject or DEFAULT_SUBJECT).strip()
    edu_level = (edu_level or "").strip()
    try:
        resolved = resolve_subject(subject_input, edu_level=edu_level, strict=True)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    crawler = await get_crawler(subject=resolved, edu_level=edu_level, strict=True)
    detail = await crawler.get_question_detail(question_id, formula_mode="svg", stem_mode="html")
    if not detail.get("success"):
        raise HTTPException(status_code=400, detail=detail.get("error") or "render_failed")

    stem_html = _sanitize_question_html(detail.get("stem_html", ""), base_url=getattr(crawler, "base_url", ""))
    return {"success": True, "question": {**detail, "stem_html": stem_html}}
